chain/network/src/concurrency/scope/mod.rs

```diff
@@ -0,0 +1,459 @@
+#![cfg(not(doctest))]
+//! Asynchronous scope.
+//!
+//! You can think of the scope as a lifetime 'env within a rust future, such that:
+//! * within 'env you can spawn subtasks, which may return an error E.
+//! * subtasks can spawn more subtasks.
+//! * 'env has special semantics: at the end of 'env all spawned substasks are awaited for
+//!    completion.
+//! * if ANY of the subtasks returns an error, all the other subtasks are GRACEFULLY cancelled.
+//!   It means that they are not just dropped, but rather they have a handle (aka Ctx) to be able
+//!   to check at any time whether they were cancelled.
+//!
+//!     let (send,recv) = channel();
+//!     ...
+//!     'env: {
+//!         spawn<'env>(async {
+//!             recv.await
+//!             Err(e)
+//!         });
+//!
+//!         while !is_cancelled() {
+//!             // do some useful async unit of work
+//!         }
+//!         // do some graceful cleanup.
+//!         Ok(())
+//!     }
+//!
+//! Since we cannot directly address lifetimes like that we simulate it via Scope and Ctx structs.
+//! Ctx is hidden in the thread_local storage and is configured by Scope.
+//! We cannot directly implement a function
+//!   run : (Scope<'env> -> (impl 'env + Future)) -> (impl 'env + Future)
+//! Because the compiler is not smart enough to deduce 'env for us.
+//! Instead we first construct Scope<'env> explicitly, therefore fixing its lifetime,
+//! and only then we pass a reference to it to another function.
+//!
+//!     let (send,recv) = channel();
+//!     ...
+//!     {
+//!         let s = Scope<'env>::new();
+//!         s.run(|s| async {
+//!             s.spawn(async {
+//!                 recv.await
+//!                 Err(e)
+//!             })
+//!
+//!             for !ctx::is_cancelled() {
+//!                 // do some useful async unit of work
+//!             }
+//!             // do some graceful cleanup.
+//!             Ok(())
+//!         }).await
+//!     }
+//!
+//! We wrap these 2 steps into a macro "run!" to hide this hack and avoid incorrect use.
+use crate::concurrency::ctx;
+use crate::concurrency::signal;
+use crate::time;
+use futures::future::{BoxFuture, Future, FutureExt};
+use std::borrow::Borrow;
+use std::sync::{Arc, Weak};
+
+#[cfg(test)]
+mod tests;
+
+// TODO(gprusak): Consider making the context implicit by moving it to a thread_local storage.
+
+struct Output<E> {
```

Comment:
This is a top-level scope ctx, which is not the same as Inner::ctx, because Inner can be service scope as well. We need references to both. Anyway, it doesn't matter now, because I've reimplemented the thing in the followup PR.